yabai: update to version 6.0.0

zack466 opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 14.0 23A344 arm64
Xcode 14.2 14C18

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

pmetzger commented:
@zack466 Do you know why the CI failed?

zack466 commented:
From what I can tell, Yabai is only officially supported on MacOS 12+ for apple silicon (ARM) and MacOS 11+ for x86-64 architectures based on their [readme](https://github.com/koekeishiya/yabai). It seems like version 5.0.9 broke compatibility with MacOS 11 on ARM (but 5.0.8 just so happened to remain compatible), which is why the build is failing. Is there a way to specify in the Portfile that Yabai does not work on MacOS 11 and below for ARM?

pmetzger commented:
If that conditional isn't exactly right, then something very like it would be. We generally would prefer to patch to make things work on earlier versions than do the known fail thing, but known fail is better than nothing (and less work for the maintainer.)

koekeishiya commented:
You can patch this part https://github.com/koekeishiya/yabai/blob/master/src/display_manager.c#L199
from
```
    uint32_t height = 0;
    SLSGetDisplayMenubarHeight(did, &height);

    bounds = CGDisplayBounds(did);
    bounds.size.height = height;
```
to
```
    bounds = CGDisplayBounds(did);
    
    uint32_t height = workspace_display_notch_height(did);
    if (height) {
        bounds.size.height = height + 5;
    } else {
        bounds.size.height = 24;
    }
```
and it will compile for macOS 11 arm64, however the scripting-addition does not support macOS 11.x.x on arm64.
I also make no guarantees that macOS 11 support won't break even further in the future on arm64.
There is a reason why it is not officially supported.

zack466 commented:
Ok, since Yabai is not officially supported on MacOS 11 on ARM and will only break further in the future (explicitly stated by the maintainer himself), I'll just use the known_fail approach suggested by @TheKevJames.

koekeishiya commented:
@zack466 Since you're doing things here atm I might add that v6.0.0 just dropped.

zack466 commented:
Um, I'm not sure why, but the CI seems to keep trying to build Yabai on ARM even though I'm specifying against it in the Portfile. It seems like the CI is running on x86_64 but is also cross-compiling for ARM (?), which fails. Also, I guess I'll just go ahead and bump the version to 6.0 (thanks for the heads up @koekeishiya).

TheKevJames commented:
Hmm everything looks right here to me. CI should be noticing the `known_fail` and not causing issues... I've never had to mark partial build arch compatibility as we need to do here, though, so perhaps there's something special which needs to be done?

pmetzger commented:
set `supported_archs` to `x86_64`.

TheKevJames commented:
@pmetzger do you know how to set `supported_archs` conditionally? We need x86 and arm for OSX 12+, but only x86 for OSX 11.

pmetzger commented:
The portfile is just a tcl script. You can put anything you want inside a conditional.

zack466 commented:
Hold on, I just realized that the line that's causing the build to fail is from the yabai makefile [here](https://github.com/koekeishiya/yabai/blob/629f6cc76057defcba4c2a0cc2dcc2a8f0e04f14/makefile#L3C24-L3C24). It seems the makefile builds for both x86_64 and arm64 regardless of platform, which is probably why changing the Portfile doesn't have an effect. @koekeishiya is there a reason for this? I would guess that with the current makefile, if Yabai doesn't compile for ARM on MacOS 11, then it won't compile for x86_64 on MacOS 11 either because of the cross-compilation.

pmetzger commented:
@zack466 This is the sort of thing patchfiles are used for in MacPorts. Obviously you should ask upstream to do something nicer of course, but you can fix the current problem if you like pretty easily.

zack466 commented:
Ok, I just added a patchfile specifically for MacOS 11 which gets rid cross-compilation to arm64 in the makefile. It also applies the same changes that `patch-makefile.diff` does (I wasn't sure if it was a good idea for two patchfiles to alter the same file in a row, so I just extracted all the changes into a new patchfile). Let me know if it looks good or if there are any changes I should make.

TheKevJames commented on sysutils/yabai/files/patch-makefile-macos11.diff:
```diff
@@ -0,0 +1,40 @@
+ 
+ all: clean-build $(BINS)
+ 
+-asan: BUILD_FLAGS=-std=c99 -Wall -g -O0 -fvisibility=hidden -fsanitize=address,undefined -mmacosx-version-min=11.0 -fno-objc-arc -arch x86_64 -arch arm64
++asan: BUILD_FLAGS=-std=c99 -Wall -g -O0 -fvisibility=hidden -fsanitize=address,undefined -mmacosx-version-min=11.0 -fno-objc-arc -arch x86_64
+ asan: clean-build $(BINS)
+ 
+-tsan: BUILD_FLAGS=-std=c99 -Wall -g -O0 -fvisibility=hidden -fsanitize=thread,undefined -mmacosx-version-min=11.0 -fno-objc-arc -arch x86_64 -arch arm64
++tsan: BUILD_FLAGS=-std=c99 -Wall -g -O0 -fvisibility=hidden -fsanitize=thread,undefined -mmacosx-version-min=11.0 -fno-objc-arc -arch x86_64
+ tsan: clean-build $(BINS)
```
I'm surprised these lines need patching, as we don't (as far as I can tell) call `make asan` or `make tsan` in the build. If we can get away with it, patching fewer things would probably be helpful in keep this easier to maintain. Of course, I could also have missed how this gets used!

If this is indeed being used, we may also need to make the `s/BUILD_FLAGS/CFLAGS/` switch here, to make sure things remain compatible with any user CFLAGS configurations.

zack466 replied:
My bad, I just deleted each instance of "arm64" from the original makefile. I've removed these lines, since it doesn't seem they require patching.

TheKevJames reviewed (approved):
Overall, I'm in favour of keeping the makefile split rather than attempting to dedupe: debugging layers of patchfiles will be difficult, and I imagine osx11 will have support dropped eventually anyway given the divergence upstream. All in all LGTM!

pmetzger commented:
Someone just let me know when we should merge this.

pmetzger commented:
Is this ready to merge?

zack466 commented:
Yeah, I think so
